cmd/skopeo/copy.go

```diff
@@ -99,24 +103,122 @@ See skopeo(1) section "IMAGE NAMES" for the expected format
+	}
+}
+
+func parseMultiArchSparse(globalOptions *globalOptions, sparseArg string) ([]imgspecv1.Platform, []digest.Digest, error) {
+	var platformList []imgspecv1.Platform
+	var instanceList []digest.Digest
+	remainder := "," + sparseArg
+	for remainder != "" {
+		parseArg := func(argStart string, argEnd string, fn func(argVal string) (bool, error)) (bool, error) {
+			if newRemainder, isArg := strings.CutPrefix(remainder, ","+argStart); isArg {
+				if !isArg {
+					return false, nil
+				}
+				var argSpec string
+				if argEnd != "" {
+					var foundArgEnd bool
+					argSpec, newRemainder, foundArgEnd = strings.Cut(newRemainder, argEnd)
+					if !foundArgEnd {
+						return false, fmt.Errorf("--multi-arch=sparse:%s: end of argument marker %s not found", argStart, argEnd)
+					}
+				}
+				handled, err := fn(argSpec)
+				if err != nil {
+					return false, err
+				}
+				if handled {
+					remainder = newRemainder
+					return true, nil
+				}
+				return false, nil
+			}
+			return false, nil
+		}
+		if isSystem, err := parseArg("system", "", func(string) (bool, error) {
+			systemPlatform := imgspecv1.Platform{
+				OS:           globalOptions.overrideOS,
+				Architecture: globalOptions.overrideArch,
+				Variant:      globalOptions.overrideVariant,
+			}
+			platformList = append(platformList, systemPlatform)
+			return true, nil
+		}); err != nil {
+			return nil, nil, err
+		} else if isSystem {
+			continue
+		}
+		if isDigest, err := parseArg("digest=[", "]", func(digestSpecList string) (bool, error) {
+			for _, digestSpec := range strings.Split(digestSpecList, ",") {
+				instanceDigest, err := digest.Parse(digestSpec)
+				if err != nil {
+					return false, fmt.Errorf("while parsing instance digest %q: %w", digestSpec, err)
+				}
+				instanceList = append(instanceList, instanceDigest)
+			}
+			return true, nil
+		}); err != nil {
+			return nil, nil, err
+		} else if isDigest {
+			continue
+		}
+		if isArch, err := parseArg("arch=[", "]", func(archSpecList string) (bool, error) {
+			wantedOS := runtime.GOOS
+			if globalOptions.overrideOS != "" {
+				wantedOS = globalOptions.overrideOS
+			}
+			for _, archSpec := range strings.Split(archSpecList, ",") {
+				p := strings.SplitN(archSpec, "/", 2)
+				if len(p) > 1 {
+					platformList = append(platformList, imgspecv1.Platform{OS: wantedOS, Architecture: p[0], Variant: p[1]})
+				} else {
+					platformList = append(platformList, imgspecv1.Platform{OS: wantedOS, Architecture: p[0]})
+				}
+			}
+			return true, nil
+		}); err != nil {
+			return nil, nil, err
+		} else if isArch {
+			continue
+		}
+		if isPlatform, err := parseArg("platform=[", "]", func(platformSpecList string) (bool, error) {
+			for _, platformSpec := range strings.Split(platformSpecList, ",") {
+				p, err := platforms.Parse(platformSpec)
```

Comment:
If we are to ever have a separate `arch=` and `platform=`, at most one of them (possibly neither) should default to `runtime.GOOS`.

Requiring the user to always provide os/arch (with an optional variant) here, using a custom parser instead of `platforms.Parse`, would for now provide a clearly deterministic feature, while still giving us the flexibility to addd that defaulting to `GOOS` in the future.